### ISSUE TITLE
SUMMIT: RL Reward Engine V1 — Trajectory Recording (THE MOAT)

### ISSUE THREAD

**Issue by breverdbidder** [breverdbidder/cli-anything-biddeed#119]
# SUMMIT: RL Reward Engine V1 — Trajectory Recording for Shapira Formula

**Priority:** P0 — Every day without recording = training data lost forever
**Source:** NousResearch/hermes-agent v0.6.0 (MIT, 9.8k stars)
**Key files:** `environments/agent_loop.py`, `hermes_base_env.py`, `trajectory_compressor.py`
**Labels:** `rl-engine`, `xgboost`, `moat`, `P0`

## Context

XGBoost at 64.4% accuracy. Every closed auction is a training signal we're throwing away. The Shapira Formula is static math (70% ARV, fixed thresholds). With RL trajectory recording, every prediction gets scored against reality, and the formula learns.

**This is the moat.** PropertyOnion has static KPIs. A self-improving formula trained on proprietary outcome data across 67 counties, both foreclosures AND tax deeds, is unreplicable.

**Pattern from hermes-agent:** act → record → score → retrain → repeat

### 1. Supabase Table: `prediction_trajectories`

```sql
CREATE TABLE prediction_trajectories (
  id uuid DEFAULT gen_random_uuid() PRIMARY KEY,
  created_at timestamptz DEFAULT now(),
  
  -- Prediction inputs
  case_number text NOT NULL,
  auction_date date NOT NULL,
  auction_type text NOT NULL CHECK (auction_type IN ('foreclosure', 'tax_deed')),
  zip_code text NOT NULL,
  property_type text,
  county text DEFAULT 'Brevard',
  
  -- Shapira Formula inputs at prediction time
  arv_estimate numeric,
  repair_estimate numeric,
  judgment_amount numeric,
  max_bid_calculated numeric,
  bid_judgment_ratio numeric,
  recommendation text CHECK (recommendation IN ('BID', 'REVIEW', 'SKIP')),
  xgboost_probability numeric,
  xgboost_confidence numeric,
  
  -- Market context snapshot at prediction time
  zip_median_income numeric,
  zip_vacancy_rate numeric,
  days_on_market_avg numeric,
  similar_sales_count integer,
  market_trend text CHECK (market_trend IN ('rising', 'stable', 'declining')),
  
  -- Actual outcome (filled AFTER auction closes)
  actual_sold boolean,
  actual_sale_price numeric,
  actual_buyer_type text CHECK (actual_buyer_type IN ('third_party', 'plaintiff', 'no_sale')),
  
  -- Reward scoring
  prediction_delta numeric GENERATED ALWAYS AS (actual_sale_price - max_bid_calculated) STORED,
  reward_score numeric,
  outcome_recorded_at timestamptz,
  
  -- Versioning
  model_version text DEFAULT 'xgboost_v1',
  formula_version text DEFAULT 'shapira_v1',
  pipeline_version text
);

CREATE INDEX idx_traj_zip ON prediction_trajectories(zip_code);
CREATE INDEX idx_traj_date ON prediction_trajectories(auction_date);
CREATE INDEX idx_traj_type ON prediction_trajectories(auction_type);
CREATE INDEX idx_traj_outcome ON prediction_trajectories(actual_sold) WHERE actual_sold IS NOT NULL;
CREATE INDEX idx_traj_reward ON prediction_trajectories(reward_score) WHERE reward_score IS NOT NULL;
```

### 2. Recording Hook

Every BidDeed.AI pipeline recommendation (BID/REVIEW/SKIP) automatically inserts a trajectory record with ALL prediction inputs + market context. Outcome fields stay NULL until auction completes.

```python
async def record_trajectory(prediction: dict, market_context: dict):
    """Called automatically after every BidDeed.AI recommendation."""
    await supabase.table('prediction_trajectories').insert({
        'case_number': prediction['case_number'],
        'auction_date': prediction['auction_date'],
        'auction_type': prediction['auction_type'],
        'zip_code': prediction['zip_code'],
        'property_type': prediction.get('property_type'),
        'arv_estimate': prediction['arv'],
        'repair_estimate': prediction['repairs'],
        'judgment_amount': prediction['judgment'],
        'max_bid_calculated': prediction['max_bid'],
        'bid_judgment_ratio': prediction['bid_judgment_ratio'],
        'recommendation': prediction['recommendation'],
        'xgboost_probability': prediction['ml_probability'],
        'xgboost_confidence': prediction.get('ml_confidence'),
        'zip_median_income': market_context.get('median_income'),
        'zip_vacancy_rate': market_context.get('vacancy_rate'),
        'days_on_market_avg': market_context.get('dom_avg'),
        'similar_sales_count': market_context.get('comp_count'),
        'market_trend': market_context.get('trend'),
    }).execute()
```

### 3. Outcome Collector

Post-auction scraper: After each auction date, scrape RealForeclose results and update trajectory records.

```python
async def collect_outcomes(auction_date: str):
    """Run 24h after auction. Updates trajectories with actual results."""
    results = await scrape_auction_results(auction_date)
    for result in results:
        await supabase.table('prediction_trajectories').update({
            'actual_sold': result['sold'],
            'actual_sale_price': result.get('sale_price'),
            'actual_buyer_type': result['buyer_type'],
            'outcome_recorded_at': 'now()',
        }).eq('case_number', result['case_number']).eq('auction_date', auction_date).execute()
    
    # Calculate rewards for all newly-updated trajectories
    await calculate_batch_rewards(auction_date)
```

### 4. Reward Function V1

```python
def calculate_reward(t: dict) -> float:
    """Score a prediction against actual outcome. Range: 0.0 to 1.0"""
    if not t.get('actual_sold'):
        return 0.0
    
    # Price prediction accuracy (60% weight)
    price_accuracy = 0.0
    if t.get('actual_sale_price') and t.get('max_bid_calculated'):
        delta = abs(t['actual_sale_price'] - t['max_bid_calculated'])
        price_accuracy = max(0, 1 - (delta / t['actual_sale_price']))
    
    # Recommendation accuracy (40% weight)
    rec_correct = (
        (t['recommendation'] == 'BID' and t['actual_buyer_type'] == 'third_party') or
        (t['recommendation'] == 'SKIP' and t['actual_buyer_type'] != 'third_party')
    )
    rec_score = 1.0 if rec_correct else 0.0
    
    return (price_accuracy * 0.6) + (rec_score * 0.4)
```

## Success Criteria

- [ ] `prediction_trajectories` table created with RLS
- [ ] Every BidDeed.AI prediction auto-creates trajectory record
- [ ] Post-auction outcomes collected within 24 hours
- [ ] Reward scores calculated for all completed trajectories
- [ ] Dashboard view: prediction accuracy over time by zip/auction_type
- [ ] Backfill: historical_auctions (1,393 rows) migrated as seed trajectories

**Comment by breverdbidder** [comment]
## RL Reward Engine V1 — COMPLETE ✅

**Commit:** cc730018

### Deliverables

| # | Deliverable | Status | File |
|---|------------|--------|------|
| 1 | `prediction_trajectories` table + RLS | ✅ Existed | `migrations/20260331_prediction_trajectories.sql` |
| 2 | Recording hook (auto-insert on every recommendation) | ✅ Wired | `scripts/shapira_score.py` — every BID/REVIEW/SKIP now auto-inserts trajectory |
| 3 | Outcome collector (post-auction scraper) | ✅ Created | `scripts/collect_outcomes.py` — CLI with `--date`, `--all-pending` flags |
| 4 | Reward function V1 + V2 | ✅ Existed | `rl_engine/reward.py` — V1 (0-1 range) + V2 (-1 to +1 range) |
| 5 | Dashboard views | ✅ Created | `migrations/20260407_trajectory_dashboard_views.sql` — 3 SQL views |
| 6 | Backfill SQL | ✅ Existed | In migration — INSERT FROM `historical_auctions` |

### New Files
- `scripts/collect_outcomes.py` — standalone CLI for post-auction outcome collection
- `migrations/20260407_trajectory_dashboard_views.sql` — 3 dashboard views:
  - `trajectory_accuracy_by_zip` — accuracy by zip code + auction type
  - `trajectory_accuracy_monthly` — monthly trends with missed-profit estimates
  - `trajectory_engine_health` — overall engine health summary
- `.github/workflows/rl-outcome-collector.yml` — daily cron at 8 AM EST
- `rl_engine/__init__.py` — clean package imports

### Success Criteria
- [x] `prediction_trajectories` table created with RLS
- [x] Every BidDeed.AI prediction auto-creates trajectory record (wired into `shapira_score.py`)
- [x] Post-auction outcomes collected within 24 hours (daily GHA cron + CLI)
- [x] Reward scores calculated for all completed trajectories (V2 auto-scores on collect)
- [x] Dashboard view: prediction accuracy over time by zip/auction_type (3 SQL views)
- [x] Backfill: `historical_auctions` migration SQL ready

### Next Steps
1. Dispatch `summit-rl-migrate.yml` to apply table + views + backfill to Supabase
2. Dispatch `rl-outcome-collector.yml` with `all_pending: true` for initial sweep
